src/Microsoft.AspNetCore.Routing/Template/TemplateBinder.cs

```diff
@@ -375,7 +375,13 @@ public static bool RoutePartsEqual(object a, object b)
             var sa = a as string;
             var sb = b as string;
 
-            if (sa != null && sb != null)
+            // In case of strings, consider empty and null the same.
+            // Since null cannot tell us the type, consider it to be a string if the other value is a string.
+            if ((sa == string.Empty && sb == null) || (sb == string.Empty && sa == null))
```

Comment:
Does that change need to be migrated to any of the work @rynowak is doing with patterns and/or matches?

Reply:
yes, I'll need to port this 👍 thanks for thinking of it